src/ompl/base/spaces/src/VanaOwenStateSpace.cpp

```diff
@@ -0,0 +1,502 @@
+        auto radiusFun = [&, this](double r)
+        {
+            update(r);
+            result.pathSZ_ = dubinsSpace_.dubins(startSZ, endSZ, result.verticalRadius_);
+            return ((result.pathXY_.length() + twopi * result.numTurns_) * r - turnS) * tanPitch + turnZ -
+                   result.deltaZ_;
+        };
+        std::uintmax_t iter = MAX_ITER;
+        try
+        {
+            auto root = boost::math::tools::toms748_solve(radiusFun, radius, 3. * radius, TOLERANCE, iter);
+            auto rval = radiusFun(.5 * (root.first + root.second));
+            if (iter >= MAX_ITER)
+                OMPL_WARN("Maximum number of iterations exceeded for high altitude Vana-Owen path");
+            return std::abs(rval) < 1e-4;
+        }
+        catch (std::domain_error& e)
+        {
+            return false;
+        }
+    }
+    else
+    {
+        // medium altitude path
+        auto zi = scratch[1];
+        auto phiFun = [&, this](double phi)
+        {
+            if (std::abs(phi)>twopi)
+                throw std::domain_error("phi too large");
+            turn(from, radius, phi, zi);
+            return (std::abs(phi) + dubinsSpace_.dubins(zi, to).length()) * radius * std::abs(tanPitch) - std::abs(result.deltaZ_);
+        };
+
+        try
+        {
+            std::uintmax_t iter = MAX_ITER;
+            result.phi_ = 0.1;
+            auto root = boost::math::tools::bracket_and_solve_root(phiFun, result.phi_, 2., true, TOLERANCE, iter);
+            result.phi_ = .5 * (root.first + root.second);
+            if (std::abs(phiFun(result.phi_)) > .01)
+                throw std::domain_error("fail");
+        }
+        catch (...)
+        {
+            try {
+                std::uintmax_t iter = MAX_ITER;
+                result.phi_ = -.1;
+                auto root = boost::math::tools::bracket_and_solve_root(phiFun, result.phi_, 2., true, TOLERANCE, iter);
+                result.phi_ = .5 * (root.first + root.second);
+            }
+            catch (...)
+            {
+                //OMPL_ERROR("this shouldn't be happening!");
+                return false;
+            }
+        }
+        result.pathXY_ = dubinsSpace_.dubins(zi, to, radius);
+        endSZ->setX((result.pathXY_.length() + result.phi_) * radius);
+        result.pathSZ_ = dubinsSpace_.dubins(startSZ, endSZ, result.verticalRadius_);
+        return std::abs(phiFun(result.phi_)) < .01 && isValid(result.pathSZ_, from);
+    }
+    return true;
+}
+
+std::optional<VanaOwenStateSpace::PathType> VanaOwenStateSpace::getPath(const State *state1, const State *state2) const
+{
+    auto from = state1->as<StateType>(), to = state2->as<StateType>();
+    double radiusMultiplier = 1.5;
+    std::array<DubinsStateSpace::StateType *, 3> scratch;
+    PathType path;
+
+    unsigned int iter = 0;
+    std::generate(scratch.begin(), scratch.end(),
+                  [this]() { return dubinsSpace_.allocState()->as<DubinsStateSpace::StateType>(); });
```

Comment:
allocation (make scratch class member and move to ctor/dtor?)

Reply:
that would violate thread safety